libr/bin/format/elf/elf.c

```diff
@@ -549,157 +530,224 @@ static void set_default_value_dynamic_info(ELFOBJ *bin) {
+	Elf_(Dyn) d = { 0 };
+	size_t i;
+	size_t j = 0;
+	size_t number_of_entries = get_maximum_number_of_dynamic_entries(dyn_size);
+
+	for (i = 0; i < number_of_entries; i++) {
+		fill_the_dynamic_entrie(bin, dyn_phdr, dyn_size, i, &d);
+
+		switch (d.d_tag) {
+		case DT_NULL:
+			break;
 		case DT_PLTRELSZ:
-			bin->dyn_info.dt_pltrelsz = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_pltrelsz = d.d_un.d_val;
 			break;
 		case DT_PLTGOT:
-			bin->dyn_info.dt_pltgot = dyn[i].d_un.d_ptr;
+			bin->dyn_info.dt_pltgot = d.d_un.d_ptr;
 			break;
 		case DT_HASH:
-			bin->dyn_info.dt_hash = dyn[i].d_un.d_ptr;
+			bin->dyn_info.dt_hash = d.d_un.d_ptr;
 			break;
 		case DT_STRTAB:
-			strtabaddr = Elf_(r_bin_elf_v2p_new) (bin, dyn[i].d_un.d_ptr);
-			bin->dyn_info.dt_strtab = dyn[i].d_un.d_ptr;
+			bin->dyn_info.dt_strtab = d.d_un.d_ptr;
 			break;
 		case DT_SYMTAB:
-			bin->dyn_info.dt_symtab = dyn[i].d_un.d_ptr;
+			bin->dyn_info.dt_symtab = d.d_un.d_ptr;
 			break;
 		case DT_RELA:
-			bin->dyn_info.dt_rela = dyn[i].d_un.d_ptr;
+			bin->dyn_info.dt_rela = d.d_un.d_ptr;
 			break;
 		case DT_RELASZ:
-			bin->dyn_info.dt_relasz = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_relasz = d.d_un.d_val;
 			break;
 		case DT_RELAENT:
-			bin->dyn_info.dt_relaent = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_relaent = d.d_un.d_val;
 			break;
 		case DT_STRSZ:
-			strsize = dyn[i].d_un.d_val;
-			bin->dyn_info.dt_strsz = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_strsz = d.d_un.d_val;
 			break;
 		case DT_SYMENT:
-			bin->dyn_info.dt_syment = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_syment = d.d_un.d_val;
 			break;
 		case DT_REL:
-			bin->dyn_info.dt_rel = dyn[i].d_un.d_ptr;
+			bin->dyn_info.dt_rel = d.d_un.d_ptr;
 			break;
 		case DT_RELSZ:
-			bin->dyn_info.dt_relsz = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_relsz = d.d_un.d_val;
 			break;
 		case DT_RELENT:
-			bin->dyn_info.dt_relent = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_relent = d.d_un.d_val;
 			break;
 		case DT_PLTREL:
-			bin->dyn_info.dt_pltrel = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_pltrel = d.d_un.d_val;
 			break;
 		case DT_JMPREL:
-			bin->dyn_info.dt_jmprel = dyn[i].d_un.d_ptr;
+			bin->dyn_info.dt_jmprel = d.d_un.d_ptr;
 			break;
 		case DT_MIPS_PLTGOT:
-			bin->dyn_info.dt_mips_pltgot = dyn[i].d_un.d_ptr;
+			bin->dyn_info.dt_mips_pltgot = d.d_un.d_ptr;
 			break;
 		case DT_BIND_NOW:
 			bin->dyn_info.dt_bind_now = true;
 			break;
 		case DT_FLAGS:
-			bin->dyn_info.dt_flags = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_flags = d.d_un.d_val;
 			break;
 		case DT_FLAGS_1:
-			bin->dyn_info.dt_flags_1 = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_flags_1 = d.d_un.d_val;
 			break;
 		case DT_RPATH:
-			bin->dyn_info.dt_rpath = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_rpath = d.d_un.d_val;
 			break;
 		case DT_RUNPATH:
-			bin->dyn_info.dt_runpath = dyn[i].d_un.d_val;
+			bin->dyn_info.dt_runpath = d.d_un.d_val;
+			break;
+		case DT_NEEDED:
```

Comment:
I think the code can be made much simpler and with less iterations over the same data if we use an `RVector` instead of a regular array. You can then use `r_vector_push` here and you don't need to pre-allocate the vector, but you can just add elements as you iterate over the dynamic tags.